macro/agent_population.py

```diff
@@ -0,0 +1,210 @@
+from dataclasses import dataclass
+from typing import NewType
+
+from HARK.core import AgentType
+from HARK.distribution import (
+    Uniform,
+    Distribution,
+    IndexDistribution,
+    combine_indep_dstns,
+)
+from xarray import DataArray
+
+ParameterDict = NewType("ParameterDict", dict)
+
+
+@dataclass
+class AgentPopulation:
+    agent_class: AgentType
+    parameter_dict: ParameterDict
+    t_age: int = None
+    agent_class_count: int = None
+
+    def __post_init__(self):
+
+        self.time_var = self.agent_class.time_vary
+        self.time_inv = self.agent_class.time_inv
+
+        self.infer_counts()
+
+    def infer_counts(self):
+
+        param_dict = self.parameter_dict
+
+        # if agent_clas_count is not specified, infer from parameters
+        if self.agent_class_count is None:
+
+            agent_class_count = 1
+            for key_param in param_dict:
+                parameter = param_dict[key_param]
+                if isinstance(parameter, DataArray) and parameter.dims[0] == "agent":
+                    agent_class_count = max(agent_class_count, parameter.shape[0])
+                elif isinstance(parameter, (Distribution, IndexDistribution)):
+                    agent_class_count = None
+                    break
+
+            self.agent_class_count = agent_class_count
+
+        if self.t_age is None:
+
+            t_age = 1
+            for key_param in param_dict:
+                parameter = param_dict[key_param]
+                if isinstance(parameter, DataArray) and parameter.dims[-1] == "age":
```

Comment:
It looks like the code here assumes that 'age' and 'agent' dimensions to the xarrays will be in a particular order.

Since the dimensions are explicitly labeled, that is not entirely intuitive -- doing it this way makes the requirement more stiffly enforced than a convention.

I wonder if there's an easy way to look up the (numeric) index of a dimension name on a DataArray